OData/src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs

```diff
@@ -185,66 +186,85 @@ private Expression CreateAggregationExpression(ParameterExpression accum, Aggreg
+                        {
+                            throw new ODataException(Error.Format(SRResources.AggregationNotSupportedForType,
+                                expression.Method, expression.Expression, propertyExpression.Type));
+                        }
+                        var sumMethod = sumGenericMethod.MakeGenericMethod(this._elementType);
+                        aggregationExpression = Expression.Call(null, sumMethod, asQuerableExpression, propertyLambda);
+
+                        // For Dynamic properties cast back to object 
+                        if (propertyAccessor.Type == typeof(object))
+                        {
+                            aggregationExpression = Expression.Convert(aggregationExpression, typeof(object));
+                        }
                     }
-                    var sumMethod = sumGenericMethod.MakeGenericMethod(this._elementType);
-                    aggregationExpression = Expression.Call(null, sumMethod, asQuerableExpression, propertyLambda);
-                }
                     break;
                 case AggregationMethod.Average:
-                {
-                    MethodInfo averageGenericMethod;
-                    if (
-                        !ExpressionHelperMethods.QueryableAverageGenerics.TryGetValue(propertyLambda.Body.Type,
-                            out averageGenericMethod))
                     {
-                        throw new ODataException(Error.Format(SRResources.AggregationNotSupportedForType,
-                            expression.Method, expression.Expression, propertyLambda.Body.Type));
+                        MethodInfo averageGenericMethod;
+                        // For Dynamic properties cast to decimal
+                        Expression propertyExpression = WrapDynamicCastIfNeeded(propertyAccessor);
+                        propertyLambda = Expression.Lambda(propertyExpression, this._lambdaParameter);
+
+                        if (
+                            !ExpressionHelperMethods.QueryableAverageGenerics.TryGetValue(propertyExpression.Type,
+                                out averageGenericMethod))
+                        {
+                            throw new ODataException(Error.Format(SRResources.AggregationNotSupportedForType,
+                                expression.Method, expression.Expression, propertyExpression.Type));
+                        }
+                        var averageMethod = averageGenericMethod.MakeGenericMethod(this._elementType);
+                        aggregationExpression = Expression.Call(null, averageMethod, asQuerableExpression, propertyLambda);
+
+                        // For Dynamic properties cast back to object 
+                        if (propertyAccessor.Type == typeof(object))
+                        {
+                            aggregationExpression = Expression.Convert(aggregationExpression, typeof(object));
+                        }
```

Comment:
Without that "Operation could destabilize the runtime" exception will be thrown. We are having ObjectProperty = DecimalValue. Per my understanding hand written up compiler will do that for us.